lib/utils/timed_counter.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"time"
+
+	"github.com/jonboulle/clockwork"
+)
+
+// TimedCounted counts events that happen over a period of time, e.g. have there been
+// more than 4 errors in the last 30 seconds. Automatically expires old events so they
+// are not included in the count.
+type TimedCounter struct {
+	clock   clockwork.Clock
+	timeout time.Duration
+	events  []time.Time
+}
+
+// TimedCounted creates a new timed counter with the specified timeout
+func NewTimedCounter(clock clockwork.Clock, timeout time.Duration) TimedCounter {
+	return TimedCounter{
+		clock:   clock,
+		timeout: timeout,
+		events:  nil,
+	}
+}
+
+// Push adds a new item into the counter, returning the current count.
+func (c *TimedCounter) Push() int {
+	c.trim()
+	c.events = append(c.events, c.clock.Now())
+	return len(c.events)
+}
+
+func (c *TimedCounter) Count() int {
+	c.trim()
+	return len(c.events)
+}
+
+func (c *TimedCounter) trim() {
+	deadline := c.clock.Now().Add(-c.timeout)
+	for len(c.events) > 0 {
+		if c.events[0].After(deadline) {
+			break
+		}
+		c.events = c.events[1:]
+	}
```

Comment:
Suggestion: Since events are naturally sorted by time, you could go back-to-front - once you find the first expired event, you can trim everything behind it as well.